icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -276,65 +311,72 @@ private PersonNamePattern getBestPattern(PersonNamePattern[] patterns, PersonNam
      * @return The name's (real or guessed) locale.
      */
     private Locale getNameLocale(PersonName name) {
-        // if the name specifies its locale, we can just return it
-        Locale nameLocale = name.getNameLocale();
-        if (nameLocale == null) {
-            // if not, we look at the characters in the name.  If their script matches the default script for the formatter's
-            // locale, we use the formatter's locale as the name's locale
-            int formatterScript = UScript.getCodeFromName(ULocale.addLikelySubtags(ULocale.forLocale(locale)).getScript());
-            String givenName = name.getFieldValue(PersonName.NameField.GIVEN, new HashSet<PersonName.FieldModifier>());
-            int nameScript = UScript.INVALID_CODE;
-            for (int i = 0; nameScript == UScript.INVALID_CODE && i < givenName.length(); i++) {
-                // the script of the name is the script of the first character in the name whose script isn't
-                // COMMON or INHERITED
-                int script = UScript.getScript(givenName.charAt(i));
-                if (script != UScript.COMMON && script != UScript.INHERITED) {
-                    nameScript = script;
-                }
-            }
-            if (formatterScript == nameScript) {
-                nameLocale = this.locale;
-            } else {
-                // if the name's script is different from the formatter's script, we use addLikelySubtags() to find the
-                // default language for the name's script and use THAT as the name's locale
-                nameLocale = new Locale(ULocale.addLikelySubtags(new ULocale("und_" + UScript.getShortName(nameScript))).getLanguage());
+        String nameScript = getNameScript(name);
+        String nameLanguage;
+        Locale specifiedNameLocale = name.getNameLocale();
+
+        Locale.Builder builder = new Locale.Builder();
+        if (specifiedNameLocale != null) {
+            builder.setLocale(specifiedNameLocale);
+            nameLanguage = specifiedNameLocale.getLanguage();
+        } else {
+            ULocale tmpLocale = ULocale.addLikelySubtags(new ULocale("und_" + nameScript));
+            nameLanguage = tmpLocale.getLanguage();
+            builder.setLanguage(nameLanguage);
+        }
+
+        // HACK to work around the fact that "Hani" isn't used in the IDs of any resource bundles
```

Comment:
There are 4 "special" scripts that we have to worry about: Jpan, Hani, Hans, Kore. I think we can lift the handling from the CLDR code.

Reply:
I just looked and didn't find anything.  Can you clue me in?  In the meantime, I've updated this to be table-driven and to take "Kore" into account.

Reply:
I was looking in the wrong place.  I'm thinking the `SpecialScripts` table in `GenerateMaximalLocales.java` might be what you're talking about?  But that looks like it does something a little different from what I'm trying to do here.  Can you expound more on what you're after here?  Is what I currently have good enough?  I'm running out of time here.

Reply:
It's not directly in the code, but here is code in the CheckPersonNames.java which is:

    public UnicodeSet getUnicodeSetForScript(String script) {
        switch (script) {
            case "Jpan":
                return JPAN;
            case "Kore":
                return KORE;
            case "Hant":
            case "Hans":
                return HANI;
            default:
                return new UnicodeSet("[\\p{sc=" + script + "}]");
        }
    }
    static final UnicodeSet HANI = new UnicodeSet("[\\p{sc=Hani}]").freeze();
    static final UnicodeSet KORE = new UnicodeSet("[\\p{sc=Hang}]").addAll(HANI).freeze();
    static final UnicodeSet JPAN =
            new UnicodeSet("[\\p{sc=Kana}\\p{sc=Hira}]").addAll(HANI).freeze();

This gets to the 'is compatible with' step, since you can get Jpan, Kore, Hant, and Hans out of maximize.

Reply:
Sorry, the "match" definition. "A few script values represent a set of scripts, such as Jpan = {Hani, Kana, Hira}. Two script codes are said to match when they are either identical, or one represents a set which contains the other, or they both represent sets which intersect. For example, Hani and Jpan match, because {Hani, Kana, Hira} contains Hani." We should probably update the spec to specifically address those 4 script values.

Reply:
Peter referred me to this same snippet, but I don't understand how it applies to this situation at all.  In my case, I've detected the script `Hani`from examining the characters and need to convert it to the default script for the relevant locale.  This code seems like it's doing the opposite.  And all I'm trying to do is fabricate something that'll let the "has resource data" check pass.

Reply:
The script-matching stuff already handles this situation, or so I believe. It's just the resource-lookup stuff I'm trying to get to do the right thing.

Reply:
Ok, great.